ml-explore/mlx
# GGUF: Avoid dequantization when format is compatible

## Proposed changes

This is a follow up for https://github.com/ml-explore/mlx/pull/350

To avoid casting to `float16`, we should unpack quantized GGUF weights to `weight, scales, biases` tensors.

Q4_1 seems like the format that more closely aligns with MLX (blocks of 32 weights with 4 bits each, and scales, and biases in fp16). Main difference is the order of the data layout.

### Manual tests

(We should add unit tests)

I tested this with this file: https://huggingface.co/TheBloke/TinyLlama-1.1B-Chat-v1.0-GGUF/discussions/3

```python
>>> import mlx.core as mx
>>> weights = mx.load("models/tinyllama-1.1b-chat-v1.0.Q4_1.gguf")
>>> mx.dequantize(weights["blk.9.attn_v.weight"], weights["blk.9.attn_v.scales"], weights["blk.9.attn_v.biases"], group_size=32)
array([[-0.0247803, -0.0568848, 0.0233154, ..., 0.00634766, 0.0279541, -0.0109253],
       [0.00671387, 0.00671387, -0.0137939, ..., -0.012207, -0.00642395, 0.0224609],
       [-0.014328, -0.018158, 0.00860596, ..., -0.0187988, -0.00262451, 0.011261],
       ...,
       [-0.00126648, 0.00540161, -0.0246124, ..., 0.000305176, 0.0160828, 0.00424194],
       [-0.00265503, -0.0170441, 0.026123, ..., 0.000640869, 0.0118408, 0.00439453],
       [0.000457764, 0.00430298, -0.0264893, ..., -0.000137329, -0.0105896, 0.00769043]], dtype=float16)
```

Expected (first 3 values of the first 3 rows and last 3 rows match):

```bash
./gguf-tools inspect-tensor ~/dev/github/mlx-examples/llms/tiny_llama_gguf/models/tinyllama-1.1b-chat-v1.0.Q4_1.gguf blk.9.attn_v.weight  | grep "\[" -A1 -B 1 | head
[
        -0.024796, -0.056885, 0.023338, 0.007294,
--
],
[
        0.006714, 0.006714, -0.013794, 0.006714,
--
],
[
        -0.014326, -0.018150, 0.008619, 0.000971,

$ ./gguf-tools inspect-tensor ~/dev/github/mlx-examples/llms/tiny_llama_gguf/models/tinyllama-1.1b-chat-v1.0.Q4_1.gguf blk.9.attn_v.weight  | grep "\[" -A1 -B 1 | tail
[
        -0.001266, 0.005405, -0.024618, 0.002069,
--
],
[
        -0.002651, -0.017040, 0.026127, -0.006248,
--
],
[
        0.000454, 0.004303, -0.026489, 0.000454,
```

I'm not sure if it would be best to leave some of the logic in gguflib

cc @awni , @antirez

Similarly, I have validated the outputs for Q4_0 and Q8_0. Both are centered at zero, instead of having custom biases per block:

```python
>>> weights = mx.load("models/tinyllama-1.1b-chat-v1.0.Q8_0.gguf")
>>> mx.dequantize(weights["blk.9.attn_v.weight"], weights["blk.9.attn_v.scales"], weights["blk.9.attn_v.biases"], group_size=32, bits=8)
array([[-0.0232849, -0.0568848, 0.0215149, ..., 0.00668335, 0.0278931, -0.0121918],
       [0.00675964, 0.0071106, -0.0128555, ..., -0.0129166, -0.00584412, 0.0224609],
       [-0.0155487, -0.0179138, 0.0100098, ..., -0.0185089, -0.00192261, 0.0109558],
       ...,
       [-0.00198364, 0.00683594, -0.0250854, ..., 0.000549316, 0.0149536, 0.00579834],
       [-0.00306702, -0.0159912, 0.026062, ..., 0.000274658, 0.0116882, 0.00415039],
       [0, 0.00540161, -0.0265808, ..., -0.00114441, -0.0119324, 0.00898743]], dtype=float16)
>>> weights = mx.load("models/tinyllama-1.1b-chat-v1.0.Q4_0.gguf")
>>> mx.dequantize(weights["blk.9.attn_v.weight"], weights["blk.9.attn_v.scales"], weights["blk.9.attn_v.biases"], group_size=32, bits=4)
array([[-0.0213318, -0.0568848, 0.0213623, ..., 0.00460815, 0.0276489, -0.0138245],
       [0.00805664, 0.00805664, -0.0134277, ..., -0.0140381, -0.00561523, 0.0224609],
       [-0.0167236, -0.0167236, 0.00836182, ..., -0.0187988, -0.00234985, 0.0117493],
       ...,
       [-0.00349426, 0.00698853, -0.0244598, ..., 0, 0.0131836, 0.00439453],
       [-0.003479, -0.017395, 0.024353, ..., 0, 0.0123596, 0.00411987],
       [0, 0.00390625, -0.0273438, ..., 0, -0.01297, 0.00778198]], dtype=float16)
```


## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [ ] I have updated the necessary documentation (if needed)
